### DISCUSSION TITLE
What is the most important missing feature for you ?

### DISCUSSION THREAD

**Discussion by fr0tt** [fr0tt/benotes#20] [General]
So I was wondering, if you, the people beside me using Benotes, have any thoughts on the subject in order for me to get (an)other perspective(s).

**Comment by AuthorShin** [comment]
In About you said "An open source self hosted notes and bookmarks taking web app. " but it's not much of a bookmarking web app specially with the absence of Exporting and Importing.

**Comment by fr0tt** [comment]
Personally I use it primarly (maybe 60%) for bookmarks. I enjoy adding bookmarks to a for me globally available service with my phone or any other device besides my own computer with basically one click.

But I have to admit I keep those online bookmarks seperated from a giant local bookmark library on my computer.
Although I obviously thought about adding import and export functionality. (There is actually already an export command but maybe that's not what you're looking for ?)

Is import and export mandatory for you when it comes to bookmarking ?
Are there other features that you think would be necessary for that experience ? (e.g. a browser extension to bookmark bookmarks)

**Reply by AuthorShin** [reply]
@fr0tt 
Honestly, I have been looking for a long time to find just one self-hosting bookmarking app that has all the basic things, and yet I haven't found any.

That's why/how I came across benotes and what I loved about it at first was tiles with link previews and the simplicity of the UI.
I know that you are working alone on this project and considering that it's in great shape and state but not enough to replace the many notepad files that I store links in them. Yeah, that's true I'm saving my links and notes on notepad files (.txt) :)

You might ask why? Well, it's safe, no one spies on them (no company no 3rd party no shit) and it's very lite in terms of size and is easy to manage and backup. You simply just copy and paste it everywhere you like.
Another reason that I'm using simple text files is that it's universally compatible with almost everything. no software, no download, no update, no not matching backup file with the new version. 
All of this happened each time I trusted a software/app (self-hosted or some closed-source one from a big company) a lot of important data has been lost due to that, so that's why right now I'm using a text file and not going to trust another app for my data unless I find it perfect and here I see the potential in benotes.

Also, you might wonder why I'm looking for a software/app for bookmarking if I'm so happy with my text files and notepad.
Well, it's because I always appreciated a good-looking UI, also I like to use one when exploring and diving into the ocean of my notes.
Another good option would be a customizable search engine, add tags and labels also categorize notes in folders.
The link preview option to see where that link would lead.
a smart importer (finding links from text files and categorizing them with the title inside folders with side notes and tags) to import my ocean of notes into small buckets.
Extra option to snapshot the current version of the links, so if the site server went offline the site is accessible even after years.
(one might want to use an API of apps like [ArchiveBox ](https://github.com/ArchiveBox/ArchiveBox) User must have the ArchiveBox too for this option).
Option to auto backup/synchronization to cloud services like the next cloud or local drive, NAS, etc.
And the most important one is to have a robust structure for preserving the data so no matter what "THE DATA IS SAFE AND SECURE"

A browser extension is nice, but I saw apps like [Linkace ](https://github.com/Kovah/LinkAce/) done the same without having to build an actual extension.
There is a link that you bookmark in your browser and each time you want to save/bookmark a website you just click on that link and a little window will pop up and ask you for confirmation on saving that website, also you can add it to lists that you have plus tags, etc.

Sorry, I didn't realize I wrote this much till the end ;D

**Reply by fr0tt** [reply]
Thanks for you detailed answer !
I can totally understand your usage of .txt files.

I added the ability to create (daily or how often you want) [backups](https://github.com/fr0tt/benotes#Backup) (as zip files) which should allow you to go back in time without issue and never have to worry about data loss. 
I felt this was a rather important feature to add but not to complicate and controversial to implement.

(On your setup you are going to need to add another volume then `benotes/storage/backup:/var/www/storage/backup`.)
(To be fair you can backup pretty much every self hosted application yourself manually. You just need to figure out what data is important or just backup everything, database(s) included.) 

Additional features (from your list) will follow but it takes time.

What you describe with Linkace sounds fascinating, but I do wonder if it is of any help (, maybe I have to try it out myself). It seems like just simply opening the application in your browser (could be bookmarked as well) and pressing create fulfills the same goal with the same amount of work/clicks ?

**Reply by AuthorShin** [reply]
Thank you very much for taking the words into action.

Yeah it is some what the same thing. 

Can you add a importer so I could import my. txt files? 
That import every url starting with https:// or http:// as link and bookmark it also add anything after `#` as comment for the following line or url of that as a comment. 
And the rest as note (the ones that don't have # at their beginning of the line or https/http)



Example: 
 
```
# Privacy orieanted search engine 
https://startpage.com
 


Meet parinaz on 19:00 at cafe
```

**Reply by fr0tt** [reply]
No problem, to be honest I quit like it myself not having to worry about data safety anymore.

I'm currently working on implementing tags and I don't have much time in next two months, but I wanted to ask you how your example should look like. The second example is easy enough, but for the first one I could think of two solutions (, but perhaps neither of them are what you would like to expect ?):

![screenshot](https://user-images.githubusercontent.com/33751346/182593082-88c7e0a8-28ff-4bf3-b5ec-d6c95788ba7c.png)

And what is the scope of one file. Would it be one collection ? Named after the file ?


Btw: While I understand your usage of .txt files, I doubt any other user uses them like that specifically, meaning this would be a highly custom solution, probably only for you ;) 
It might make sense to split it into two stages: a general importer parsing .html (like exported bookmarks from most browsers) or .json files and a script that transforms your .txt files into the former.

**Reply by AuthorShin** [reply]
The middle one looks good but: 1. Can you add snapshot to it so it will show a screenshot of the website instead of just logo (icon). 2. in the middle one what happened to "Privacy oriented search engine" Where does that shown?

Yeah it can be collection too, an option when importing to add as a collection or not would be nice.
You are right it's looks like a V.I.P. option :D

**Reply by fr0tt** [reply]
1. I was actually researching that after your first message in this thread, but it would require to extend the current stack of this project with an additional programming language which makes it more difficult to host "regulary" (and increases the workload for the additional service needed when using docker-compose), so I'm still thinking about it. I'm also not sure so far how well it actually works/looks for large or picture heavy websites.

2. I was afraid you are going to ask that. It's nowhere. The description is automatically taken from the website meta data. Currently there is no way to change it manually, design-wise.

Where should it be stored if not in a collection? In 'uncategorized' ?

**Reply by AuthorShin** [reply]
1. What do you mean by "large or picture heavy websites"? it's not gonna be heavy or have any load at all except the moment you add the note/link. because it's just once and after that it will only loading that image from storage. For example take a look at [gtmetrix.com](https://gtmetrix.com) when you enter a website address it will showing you a image (screenshot of that) which got the path of 'https://gtmetrix.com/job-status/Wv8KduVf/screenshot.jpg'.
2. Why not just make it hover able that when you hover the tile with mouse cursor it will show you that note (# one) also make it editable within an option in side those 3dots button.

There can be a general category (collection) for it or an option to set the default collection for imported notes.

**Reply by fr0tt** [reply]
1. Well I thought you mean a screenshot from the entire website. 
Your example only shows the upper part in a bad resolution, which is in imho often not better than the social media thumbnail already provided.

2. This wouldn't be exactly a well reprehensible design because it would be unexpected, not to mention that there already other information you can see while hovering (entire URL, post URL).
The three-dots-menu would be an easy solution but I would still be a little weird, since it's the only data that you could change this way. 
To be honest I was always a little unhappy that - in contrast to the text items - link items are not "what you see is what you get". Perhaps I should split the entire edit/create page into a text and link page and you have to choose beforehand what you want to create.

**Comment by godsgood33** [comment]
Somebody posted this in r/selfhosted, so I wanted to try it out.  I think it serves as a really nice home page for browsers.  I've used "Speed Dial 2" https://speeddial2.com plugin for Chrome based browsers for a long time and have found it extremely helpful.  They recently announced they are changing how they are doing things and are going to start requiring a subscription.  So I'd love to see some UI changes that are a little closer to that (moving collections to a top nav, ability to select, upload, or link an image for the bookmark, and ability to remove the text just leaving a title).  Theming, background images, and customized tiles would be icing on the cake.  I can provide some screenshots if it would be helpful.

**Reply by fr0tt** [reply]
Thanks for your post ! 

Moving collections to the top instead of the sidebar seems not ideal because it seems to be designed for a small amount of them (as far as I can tell by looking at their screenshots) instead of 5 to let's say 20 or 50 (in the future even nested).

Thumbnails are directly used from the website if the website specifies one. However this seems like a good addition if that's not the case.
When you write removing text from bookmarks do you mean on a per bookmark basis and in any case do you mean the thumbnail should expand then ?

What do you mean by customized tiles ? As background images ?

**Comment by godsgood33** [comment]
First, let me say, I think this project is great!  You've done a lot!  All of what I mention is simply sharing my use case for how I would like to use it.  Since I know PHP and it looks like you are using Laravel, I might fork it and see about making some of the changes myself as a suggestion.

You're right, the idea of "collection" does seem to imply more, so that's probably not the best, but I propose a couple options below.

Yes, Speed Dial can use screen grabs from the URL as well, or you can specify an image.  They also use the favicon if the site uses one.  I prefer to specify an image because I don't like my images changing if the favicon or screen grab changes.  And some large services like Google, for example, if I just used the site image, it would just be the Google logo for all of their different services.

What I mean when I say getting rid of text is I don't want to see the meta description for a site.  If I just put in a URL I just want a shortcut to that page with no text.  This is maybe where splitting the different types of tiles may be helpful.  "Add note", "Add bookmark" and bookmarks only show the image, title, and maybe the URL.
![Benotes](https://user-images.githubusercontent.com/1922710/204329746-cb5fe181-89c6-4d45-a847-5a8fda51f9b0.png)

## Big Picture
This is the main page as it appears on Speed Dial when opening a new tab.  I could see setting my browsers home page to Benotes, so that it would open something like this all the time.  Maybe you could allow setting a collection type?  Something like it would give options for "Nav", "Notes", "Both" and only collections of type "Nav" would appear in the top navigation.  Or maybe make a separate thing altogether.  Speed Dial calls them "Groups".  So groups are available in the top nav, and make them a "superset" of collection, so they can contain collections like you were saying.
![Big picture](https://user-images.githubusercontent.com/1922710/204321429-2ee27600-02e1-4910-a8c5-b348b3fcfca4.png)

## Bookmarks
This shows what options are available for the bookmarks
![Settings](https://user-images.githubusercontent.com/1922710/204321424-686b06e9-9c69-4163-bd4a-fdbac19f0817.png)

## General Preferences
This shows the spacing on the page that I was referring to.  Number of Columns (3-20, but I like 6), self-explanatory.  Spacing = num of pixels between tiles, both horizontal and vertical.  Width = num of pixels to take up on the screen, probably uses a range of 60-99% of available pixels)
![Spacing](https://user-images.githubusercontent.com/1922710/204321426-2ce612b5-3aa3-46ed-9b7d-4e0422a73999.png)

## Theming Options
This just the theme options, you can specify an image, positioning, repeating, etc.  Since you can use this with docker, what about looking for a "background.ext" file in a directory and it will automatically make that the background, then within docker, have a shared directory that on the host machine we can copy the desired file to and boom!  
![Theme](https://user-images.githubusercontent.com/1922710/204321428-3b93ec89-411e-414c-9ea9-51a377f87965.png)

## Groups
This is another example of how you might be able to do the top nav...if it is enabled in a setting like this, then it appears.
![Groups](https://user-images.githubusercontent.com/1922710/204321434-e3ef0663-b0a2-47f9-af2b-f65837cfb1e0.png)

**Reply by fr0tt** [reply]
Thank you very much. That is very kind of you to say !
Yes indeed is uses "Laravel's" Lumen (which I will migrated to Laravel in the future).

While it is indeed a very good question to challenge the "Create note/boomark" design, it has nothing to do with the description. 
The description only exists for bookmarks and I think it is quit useful when you e.g. have a couple of posts let's say from Medium and an image and title are not enough to help you remember what each is about.
However I could think of two solutions for this. First by introducing a toggle in (a newly created) Preferences which allows you to globally define if descriptions should be used or not. Or the other solution might be to decide if a description should be visible per bookmark. So for every single boomark you get to decide if the description should be hidden.

To be honest I am not so sure about the group concept for this application because as I said before visibility can be poor if you have a lot of them. (After all people don't just use this app for some bookmarks but rather maybe for a couple of hundreds with dozens of (root) collections). Not to mention it would be a poor experience on mobile.

Customizing thumbnails is indeed a great idea! This would however lead to some infrastructure problems. So far this app can be used without any permanent storage (except the database) and (what's far worse) screenshosts could not be created without Nodejs and a headless browser (which would drastically increase requirements for both admins and servers). That is actually the reason why I haven't already implement something like that.

The idea with the number of items per "row" seems interesting.

Theming described as in your screenshot is not really a priorty right now. However I think could see an option to add custom css stylesheets to achieve the same goal.

**Comment by MohamedElashri** [comment]
One of the things that is easy to implement to ease of use in to add a bookmarklet that use can add using its instance URL. Something like `linkding` do it. I created it for instance and it is working like a charm (no need to a full browser extension for now)

The code is simple and could be something like 


```
javascript: (function() {    var bookmarkUrl = window.location;    var applicationUrl = '$APP_URL/c/0/p/create';    applicationUrl += '?url=' + encodeURIComponent(bookmarkUrl);    applicationUrl += '&auto_close';    window.open(applicationUrl);})();
```

**Reply by fr0tt** [reply]
That's actually pretty cool !

**Reply by fr0tt** [reply]
I'm definitely going to add that !

**Comment by vmsman** [comment]
I have scoured the code and the project appears complete.  I installed via docker-compose and the product gives me a login screen.  Nothing in the .env file, the source code or the documentation points to how to create initial login credentials or what default login credentials are.  I am assuming this is a closed beta by invitation only.  I will check back in a few months.

**Reply by MohamedElashri** [reply]
> I have scoured the code and the project appears complete. I installed via docker-compose and the product gives me a login screen. Nothing in the .env file, the source code or the documentation points to how to create initial login credentials or what default login credentials are. I am assuming this is a closed beta by invitation only. I will check back in a few months.

Have you followed all the steps after `docker-compose up -d` on this [page](https://benotes.org/docs/installation/docker-compose), because this will help you with creating user?

**Reply by fr0tt** [reply]
As @MohamedElashri pointed out correctly you can find all necessary information here on the GitHub repository readme or on the website. Everyone is allowed to use it at it's current state without an invitation or limitation.

**Comment by vmsman** [comment]
@MohamedElashri  In fact, I did follow the two steps listed on the page after the docker-compose up -d.  I have continually received an error which I have to assume is unimportant because it is documented nowhere else on the site.

`scott@Benotes:~$ docker-compose exec --user application app sh
/var/www $
/var/www $ php artisan install
Initiate installation...


 0/4 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]   0%


   ErrorException

  file_put_contents(/var/www/.env): Failed to open stream: Permission denied

  at vendor/laravel/framework/src/Illuminate/Foundation/Console/KeyGenerateCommand.php:96
     92▕     {
     93▕         file_put_contents($this->laravel->environmentFilePath(), preg_replace(
     94▕             $this->keyReplacementPattern(),
     95▕             'APP_KEY='.$key,
  ➜  96▕             file_get_contents($this->laravel->environmentFilePath())
     97▕         ));
     98▕     }
     99▕
    100▕     /**

      +13 vendor frames
  14  app/Console/Commands/InstallCommand.php:47
      Illuminate\Console\Command::call("key:generate")

      +13 vendor frames
  28  artisan:37
      Illuminate\Foundation\Console\Kernel::handle(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
`

**Reply by MohamedElashri** [reply]
This problem arises from the fact that the user you are using doesn't have ownership of the directory you have `.env` file there. There was a discussion about that in this [issue](https://github.com/fr0tt/benotes/issues/53). You will find the solution also there in my comments.

**Comment by ArbiterGR** [comment]
Hello! a feature that i'm missing, while not a core one but important to me nonetheless, would be to be able to use my own identity provider for authentication, either via http_auth remote header (eg x-remote-user  header) or openid. This would definetly make the instance more secure and less of a risk having it exposed to the internet.

**Comment by centrdev** [comment]
would love to see collection "folders" for better organizing of collections
something like:

Folder1
 - My Collection 1
 - My Collection 2
Folder2
 - My Collection 3
 - My Collection 4

**Reply by fr0tt** [reply]
Thanks for your suggestion.

![1682062415432.jpg](https://user-images.githubusercontent.com/33751346/233572708-2d6f4e6a-4fa6-4a09-8f80-374130756690.jpg)

Just to be clear: You don't mean something like this ? Because this is already possible.

**Reply by centrdev** [reply]
lol, yes, that's what i mean 🫢

how do i do that ?

EDIT:

NVM, just had to update, all good ✌️

**Reply by fr0tt** [reply]
When you create or edit a collection there's a field directly under 'name' called 'subcollection of' ;) 
That is if your instance is more or less up-to-date.

But I'm curious now: Did you not see this before or do you think this was meant for something entirely else ? Or do you just use an older version of Benotes ?

**Reply by centrdev** [reply]
yeah, sorry, i was on an older version. thought i had done a docker install but did it the old fashioned way so had to manually update.

**Comment by ShyViolets** [comment]
I was wondering if moodboard format and other presentation formats were considered?
i just stumbled across this and it looks really promising (I use raindrop currently) - but I wasn't able to find a demo to check out the presentation formats.

**Comment by philipp-koch** [comment]
I'm a little hesitant to fully "invest" in using Benotes as long as I can't get my data out easily if necessary. I'm currently looking for an [Obsidian](https://obsidian.md/) alternative (and hope that Benotes might be it). One thing I like about Obsidian is that it saves all "notes" data as markdown files (on a local drive - it's not designed as a web application, although it can be run in the browser). It's obviously still challenging to keep all the references between the markdown files intact (Obsidian relies on filenames, but tracks links/backlinks and propagates any changes throughout all files), once one plans to use them elsewhere, but at least they're accessible in a non-proprietary way.

So, to answer your initial question, for me the three most important features would be:
# 1. **Easy export of all notes** 
  * Offer an export of all notes to a zip archive that can be downloaded. This would contain folders (following the collections hierarchy) with all the notes as markdown files in their respective folder.
  * Metadata (tags, backlinks etc.) of each note could either be generated in a dedicated "metadata-block" at the top or bottom of the markdown file, or an accompanying sidecar-file in an accessible, structured format (YAML, markdown or maybe XML, JSON, CSV)
  * Import of a single file exported in that manner or a bunch of them would complete the feature.
  
# 2. "Managed" **linking between notes**, i.e. 
  * Benotes would show the correct title even if it has changed after linking to the note
  * Manage Backlinks (this note is references by the following notes: a, b, c) and keep them up-to-date

# 3. Allow **images in notes**, and they could be added
  * via drag and drop from a local source or another browser tab 
  * pasted from the clipboard (e.g. for screenshots)
  * via "image catalog" (list of all images currently saved within Benotes)
  * Images would be saved automatically when added and also automatically be discarded of, once they're no longer used anywhere. 

# Some other ideas
That being said, I think there's a lot of unreleased potential in the search function:
  * It should not only search full-text, but also metadata (e.g. if I type in a tag or maybe explicitly ``#tag`` (with the hashtag), notes that are tagged with it should be found as well). Might be an option that could be turned on or off ("also search in tags" or similar) (also see #76)
  * It should link search tokens in a OR-way: "github benotes" should find anything that contains either "github", "benotes" or both; right now, it finds only notes that contain exactly the given search-term (verbatim and case-sensitive).
  * Search offers collections as filter (I think? As in "only search with this collection"?), but only the to-most ones, not their sub-collections
  * Search should be case-insensitive (right now it doesn't find "Berlin" if I search for "berlin")
  * A click on a collection should show all notes in it (also those in its sub-collections), because right bow I need to know beforehand where - in which (sub-)collection - I put something in order to find it. Besides "UNcategorized" there should also be an "all notes" category/button
  * Instead of a separate Search menu button, it would be so much more powerful to have the search simply start by typing when viewing a number of notes (sort of like [dashy](https://demo.dashy.to/#) does, for instance)
 
If it helps, I could make separate issues out of these (and other) enhancement ideas so that they can be tracked and discussed more easily.

**Reply by fr0tt** [reply]
I addressed some of your search problems in the newest version 2.7.1

**Reply by philipp-koch** [reply]
While I have taken a look at the updated version, I can't seem to spot the differences in search behavior other than now tags seem to be included in the search, and this seems to be case-insensitive and match also partial terms ("ber" or "Ber" would both find items tagged "berlin")?

**Reply by fr0tt** [reply]
Well it's only a minor release that solves a couple of things, among them are: searching in subcollection automatically as well, searching in subcollection at all and fix searching with PostgreSQL.

**Comment by vmsman** [comment]
To me, the number one missing feature is the ability to paste a graphic image into a note.

**Reply by fr0tt** [reply]
You mean uploading them ? 
Because you can already use an image from another website by copying it.

**Reply by vmsman** [reply]
I have found no way to make a screenshot image a part of a note.  For example, in Joplin notes I have both the options to drag and drop an image and also to click an icon on the note editor to open the local file manager and upload an image that would become an inline image in the note.

**Reply by fr0tt** [reply]
Just wanted to be sure what you mean

**Reply by fr0tt** [reply]
I include it in new v2.8.0-beta release.

**Reply by vmsman** [reply]
Awesome!  I will watch for it.  Thanks so much.

**Comment by rldaugherty** [comment]
The biggest thing for me would to have a better UI experience. The linking and everything works great for me, but I wish that there was a "Landing" page (ie: homepage) that I could show maybe the most recent 5 videos, or a few from specific categories. When I add a bookmark, I'm afraid I would forget about it, and without some way of seeing it on a homepage, it would just get lost in the void.

Maybe something to the effect of:

**MOST RECENT**
(Lists the 5 most recent links, note, etc.)

**VIDEOS**
(Lists the 5 favorited links that I specified in the "VIDEOS" Category)

**LINKS**
(Lists 5 links that I favorited.)

And so on.
